### PR TITLE
Change coinbase to exclude

### PR DESCRIPTION
**Pull request by KaustubhShamshery**
## Description
The guard mode nodes, will not be awarded rewards and skipped during coinbase.

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [x] local machine test

- [x] small-scale cloud test

### CODE REVIEW

**Comment by bb111189** [comment]
Will this include the lucky draw rewards too?

File: src/libDirectoryService/Coinbase.cpp [hunk 1]
```diff
@@ -333,7 +340,16 @@ void DirectoryService::InitCoinbase() {
   for (const auto& shard : m_coinbaseRewardees[m_mediator.m_currentEpochNum]) {
     if (count == shardIndex) {
       uint16_t rdm_index = lastBlockHash % shard.second.size();
-      const Address& winnerAddr = shard.second[rdm_index];
+      if (GUARD_MODE) {
+        while (
+            Guard::GetInstance().IsNodeInDSGuardList(shard.second[rdm_index]) ||
```
Comment (bb111189) [inline src/libDirectoryService/Coinbase.cpp]: use at() instead of []

File: src/libDirectoryService/Coinbase.cpp [hunk 2]
```diff
@@ -333,7 +340,16 @@ void DirectoryService::InitCoinbase() {
...
+            Guard::GetInstance().IsNodeInShardGuardList(
+                shard.second[rdm_index])) {
```
Comment (bb111189) [inline src/libDirectoryService/Coinbase.cpp]: use at() instead of []

File: src/libDirectoryService/Coinbase.cpp [hunk 3]
```diff
@@ -333,7 +340,16 @@ void DirectoryService::InitCoinbase() {
...
+          rdm_index = (rdm_index + 1) % shard.second.size();
+        }
+      }
+      const Address& winnerAddr =
+          Account::GetAddressFromPublicKey(shard.second[rdm_index]);
```
Comment (bb111189) [inline src/libDirectoryService/Coinbase.cpp]: use at() instead of []

File: src/libDirectoryService/Coinbase.cpp [hunk 4]
```diff
@@ -71,15 +72,13 @@ bool DirectoryService::SaveCoinbaseCore(const vector<bool>& b1,
...
-          Account::GetAddressFromPublicKey(pubKey));
+      m_coinbaseRewardees[epochNum][shard_id].push_back(pubKey);
       if (m_mapNodeReputation[pubKey] < MAX_REPUTATION) {
         ++m_mapNodeReputation[pubKey];
       }
     }
     if (b2.at(i)) {
-      m_coinbaseRewardees[epochNum][shard_id].push_back(
-          Account::GetAddressFromPublicKey(pubKey));
+      m_coinbaseRewardees[epochNum][shard_id].push_back(pubKey);
```
Comment (bb111189) [inline src/libDirectoryService/Coinbase.cpp]: use at() instead of []

Comment (KaustubhShamshery) [reply src/libDirectoryService/Coinbase.cpp]: throws exception if first time.

File: src/libDirectoryService/Coinbase.cpp (hunk 3 above)
Comment (KaustubhShamshery) [reply src/libDirectoryService/Coinbase.cpp]: OK

File: src/libDirectoryService/Coinbase.cpp (hunk 1 above)
Comment (KaustubhShamshery) [reply src/libDirectoryService/Coinbase.cpp]: OK

File: src/libDirectoryService/Coinbase.cpp (hunk 2 above)
Comment (KaustubhShamshery) [reply src/libDirectoryService/Coinbase.cpp]: OK

**Comment by bb111189** [comment]
What is the status of local and small scale test?

File: src/libDirectoryService/Coinbase.cpp [hunk 5]
```diff
@@ -392,7 +413,16 @@ void DirectoryService::InitCoinbase() {
   for (const auto& shard : m_coinbaseRewardees[m_mediator.m_currentEpochNum]) {
     if (count == shardIndex) {
       uint16_t rdm_index = lastBlockHash % shard.second.size();
-      const Address& winnerAddr = shard.second[rdm_index];
+      if (GUARD_MODE) {
+        while (Guard::GetInstance().IsNodeInDSGuardList(
+                   shard.second.at(rdm_index)) ||
+               Guard::GetInstance().IsNodeInShardGuardList(
+                   shard.second.at(rdm_index))) {
+          rdm_index = (rdm_index + 1) % shard.second.size();
```
Comment (bb111189) [inline src/libDirectoryService/Coinbase.cpp]: Possible to explain what this line is for?

**Pull request merged**